src/core/common.rs

```diff
@@ -173,12 +208,83 @@ pub(crate) fn dumps(ked: &Value, kind: Option<&str>) -> Result<Vec<u8>> {
     }
 }
 
+pub(crate) fn sniff(raw: &[u8]) -> Result<SniffResult> {
+    lazy_static! {
+        static ref REVER: Regex = Regex::new(REVER_STRING).unwrap();
+    };
+
+    if raw.len() < MINIMUM_SNIFF_SIZE {
+        return err!(Error::Value(format!(
+            "need more bytes than {bytes} to sniff",
+            bytes = raw.len()
+        )));
+    }
+
+    let raw = &String::from_utf8(raw.to_vec())?;
+    let start = match REVER.shortest_match(raw) {
+        Some(m) => m - VERSION_FULL_SIZE,
+        None => return err!(Error::Value(format!("invalid version string in raw = {raw:?}"))),
+    };
+
+    if start > MAXIMUM_START_SIZE {
+        return err!(Error::Value(format!(
+            "invalid version string in raw = {raw:?} start = {start}"
+        )));
+    }
+
+    let vs = &raw[start..(start + VERSION_FULL_SIZE)];
+
+    let ident = REVER.replace_all(vs, "$ident").to_string();
+    let major = u8::from_str_radix(&REVER.replace_all(vs, "$major"), 16)?;
+    let minor = u8::from_str_radix(&REVER.replace_all(vs, "$minor"), 16)?;
+    let kind = REVER.replace_all(vs, "$kind").to_string();
+    let size = u32::from_str_radix(&REVER.replace_all(vs, "$size"), 16)?;
+    let version = Version { major, minor };
+
+    if !SERIALS.contains(&kind.as_str()) {
+        return err!(Error::Validation(format!("invalid serialization kind {kind}")));
+    }
+
+    Ok(SniffResult { ident, kind, version, size })
+}
+
 #[cfg(test)]
 mod test {
     use crate::core::common;
-    use crate::data::data;
+    use crate::data::{data, Data};
     use rstest::rstest;
 
+    #[test]
+    fn loads() {
+        let raw = &data!({}).to_json().unwrap().as_bytes().to_vec();
+        assert!(common::loads(raw, None, None).is_ok());
+    }
+
+    #[test]
+    fn sniff_sad_paths() {
+        assert!(common::sniff(&[]).is_err()); // minimum 29 octets
+        assert!(common::sniff(
+            &data!({"v":"version string must be valid!"}).to_json().unwrap().as_bytes()
+        )
+        .is_err());
+        assert!(common::sniff(
+            &data!({"i":"needs to start within 12 characters!","v":"KERI10JSON000000_"})
+                .to_json()
+                .unwrap()
+                .as_bytes()
+        )
+        .is_err());
+        assert!(common::sniff(&data!({"v":"KERI10ABCD000000_","confusing but necessary filler":"hmm...maybe a 12 octet magic prefix?"}).to_json().unwrap().as_bytes()).is_err());
```

Comment:
I didn't understand why this was necessary.. why the 12 extra bytes? Haven't read the paper in a while